### PR TITLE
PROD-229 fixed bug for wrong open range products

### PR DESCRIPTION
**Pull request by markkampstra**
Some products are marked as 'open_range = 1' while they are fixed
denomination products. An extra check is added to determine if an open
range product is actually a fixed denomination product.

All rspec tests work with prerecorded API responses (VCR)

Updated .travis.yml to run rspec.

### CODE REVIEW

File: spec/integration/api_spec.rb [hunk 1]
```diff
@@ -44,126 +35,175 @@
 
   context '#pricelist' do
     it 'should return a list of countries' do
-      pricelist = TransferToApi::Pricelist.get_countries
-      expect(pricelist.error_code).to eq 0
-      expect(pricelist.countries.length).to be > 10
-
-      country = pricelist.countries.detect { |n| n.id == @open_range_country_id}
-      expect(country.name).to eq @open_range_country_name
+      VCR.use_cassette('countries') do
```
Comment (DavyCardinaal) [inline spec/integration/api_spec.rb]: Awesome ⏺  🎥

File: spec/integration/api_spec.rb [hunk 2]
```diff
@@ -44,126 +35,175 @@
...
+        expect(pricelist.products.length).to be > 3
+
+        product = pricelist.products.first
+        expect(product.class).to eq TransferToApi::PricelistFixedProduct
+        expect(product.operator.country.id).to eq @fixed_denomination_country_id
+        expect(product.operator.name).to eq @fixed_denomination_operator_name
+      end
+    end
+
+    it 'detects a wrongly marked open range product and treats it like a fixed denomination product' do
```
Comment (DavyCardinaal) [inline spec/integration/api_spec.rb]: Yeah! 🌹

File: spec/integration/api_spec.rb [hunk 3]
```diff
@@ -44,126 +35,175 @@
...
-      expect(topup.class).to eq TransferToApi::TopupPin
-      expect(topup.pin_based).to eq 'yes'
-      expect(topup.pin_code).to match(/^[0-9]{4}[0-9]*$/)
+      info = nil
+      VCR.use_cassette('pin-based-msisdn-info') do
+        info = TransferToApi::MsisdnInfo.get(@pin_based_phone_number)
+        expect(info.error_code).to eq 0
+      end
+
+
```
Comment (DavyCardinaal) [inline spec/integration/api_spec.rb]: I think we can remove the second empty line

File: spec/integration/api_spec.rb [hunk 4]
```diff
@@ -44,126 +35,175 @@
...
-      expect(topup.class).to eq TransferToApi::Topup
-      expect { topup.pin_based }.to raise_error(NoMethodError)
-      expect { topup.pin_code }.to raise_error(NoMethodError)
+      info = nil
+      VCR.use_cassette('fixed-denomination-msisdn-info') do
+        info = TransferToApi::MsisdnInfo.get(@fixed_denomination_phone_number)
+        expect(info.error_code).to eq 0
+      end
+
+
```
Comment (DavyCardinaal) [inline spec/integration/api_spec.rb]: Same here, double empty line.

File: .travis.yml [hunk 5]
```diff
@@ -1,3 +1,12 @@
 language: ruby
-rvm:
-  - 2.2
+sudo: required
+rvm: 2.2.2
+
```
[the remark this answers is not in the document]
Comment (markkampstra) [reply .travis.yml]: ✅

File: lib/transfer_to/pricelist.rb [hunk 6]
```diff
@@ -109,6 +109,13 @@ def self.get_products_for_operator_from_client(client, operator_id)
       pricelist
     end
 
+    def self.open_range?(data)
+      # The TransferTo API sometimes returns an open_range product, when it's actually a fixed range product
+      # see: https://creativegroupdev.atlassian.net/browse/PROD-229
+      multiple_products = !!(data[:product_list] =~ /,/)
+      data[:open_range] == '1' && !multiple_products
```
[the remark this answers is not in the document]
Comment (markkampstra) [reply lib/transfer_to/pricelist.rb]: You are right, exclude? is much cleaner than ❗️ ❗️ ❗️

File: spec/integration/api_spec.rb [hunk 7]
```diff
@@ -237,27 +282,32 @@ def test_error(error_phone_number, expected_exception_code)
...
+        expect(id_from_key.transaction_id).to eq @pin_based_transaction_id
+      end
     end
   end
 
   context '#check_wallet' do
     it 'should get the credit in the transferto account' do
-      credit = TransferToApi::CheckWallet.get
+      VCR.use_cassette('check-wallet') do
+        credit = TransferToApi::CheckWallet.get
```
Comment (Tuxified) [inline spec/integration/api_spec.rb]: would be nice to add an assert, for example `expect(credit).to be_a Credit` or similar?

File: spec/integration/api_spec.rb [hunk 8]
```diff
@@ -44,126 +35,173 @@
...
+          wholesale_price_list: '5,6',
+          skuid_list: '123,456',
+          open_range: '1'
+        }
+      )
+      client = TransferToApi::Client.new('user', 'pass')
+      allow(client).to receive(:run_action) { wrong_open_range_response }
+      pricelist = TransferToApi::Pricelist.get_products_for_operator_from_client(client, 'operator1')
+      expect(pricelist.products.size).to eq 2
+      expect(pricelist.products.first.class).to eq TransferToApi::PricelistFixedProduct
```
Comment (Tuxified) [inline spec/integration/api_spec.rb]: You can also write it as: `expect(pricelist.products.first).to be_a TransferToApi::PricelistFixedProduct`

File: spec/integration/api_spec.rb (hunk 2 above)
Comment (Tuxified) [reply spec/integration/api_spec.rb]: Squash all the bugs!🤘

**Pull request merged**